backend/tracim_backend/views/core_api/user_controller.py

```diff
@@ -320,6 +333,10 @@ def set_user_email(self, context, request: TracimRequest, hapic_data=None):
         uapi = UserApi(
             current_user=request.current_user, session=request.dbsession, config=app_config  # User
         )
+        if not has_write_rights("email", app_config, request):
+            return HTTPBadRequest(
```

Comment:
You should raise a `ReadOnlyFieldException` (which you have to create) + associated error code.

For some example, search `WORKSPACE_NOT_FOUND` in `error.py` file. We can have a talk tomorrow morning

Reply:
The way to implement errors in the API is the following one:

```
uapi = UserApi(current_user=request.current_user, session=request.dbsession, config=app_config)
if has_write_rights("email", app_config, request):
    user = uapi.set_email(
        request.candidate_user,
        hapic_data.body.loggedin_user_password,
if not has_write_rights("username", app_config, request):
    [...]
```

and the implementation of the check should be the following one:

```
def has_write_rights(field: str, app_config: CFG, request: TracimRequest) -> bool:
    """
    Check if user can write provided field (is admin or field is writable)
    """
    read_only_fields = app_config.USER__READ_ONLY_FIELDS.get(request.candidate_user.auth_type.value)
    if request.current_user.profile.name != "ADMIN":
        if read_only_fields is not None:
            if field in read_only_fields:
                raise ReadOnlyFieldException(f"You can't modify the field {field}")
    return True
```

The explaination is the following one:

- the function will raise an exception
- the exception would be caught by hapic using the following decorator: `@hapic.handle_exception(ReadOnlyFieldException, HTTPStatus.BAD_REQUEST)`
- hapic will process the exception with associated code and message and builds the right JSON struct for frontend being able to process errors

Reply:
Look at the rout signature:

```
    @hapic.with_api_doc(tags=[SWAGGER_TAG__USER_ENDPOINTS])
    @hapic.handle_exception(WrongUserPassword, HTTPStatus.FORBIDDEN)
    @hapic.handle_exception(EmailAlreadyExists, HTTPStatus.BAD_REQUEST)
    @hapic.handle_exception(ExternalAuthUserEmailModificationDisallowed, HTTPStatus.BAD_REQUEST)
    @check_right(has_personal_access)
    @hapic.input_body(SetEmailSchema())
    @hapic.input_path(UserIdPathSchema())
    @hapic.output_body(UserSchema())
    def set_user_email(self, context, request: TracimRequest, hapic_data=None):
        [...]
```

The `WrongUserPassword`, `EmailAlreadyExists`, `ExternalAuthUserEmailModificationDisallowed` errors are catched and transform to HTTP Bad Request with tracim standard JSON error  

You may probably even implement something similar to `has_personal_access`

Reply:
The reason for this software design is:

- to keep function as readable as possible
- to manage errors a standard way
- to manage errors out-of-code-path
- to ensure we'll always return valid JSON (every HTTP answer is processed using hapic

Code is easier to read and not less robust

Reply:
Resolved in last commit. Thanks for the explanation ! It is indeed way better to read / code